sdk/python/kfp/dsl/_pipeline.py

```diff
@@ -127,10 +127,17 @@ def __enter__(self):
       raise Exception('Nested pipelines are not allowed.')
 
     Pipeline._default_pipeline = self
+
+    def register_op_and_generate_id(op):
+      return self.add_op(op, op.is_exit_handler)
```

Comment:
This is not generating a distinct ID for the containerOp. 
since this function will replace the _make_hash_based_id_for_container_op, how is the distinct ID generated?

Reply:
It does generate a distinct ID. It just does that differently.
The default implementation generates unique IDs, but they're different each time (compiler tests won't like that). (A solution might be to use proper hashing.)

The replacement function also generates a unique ID - it adds indices whenever the task with the same name already exists:

https://github.com/kubeflow/pipelines/blob/339b2dc8a10c062c3d58926a5d98c1ec9a825b33/sdk/python/kfp/dsl/_pipeline.py#L153

The idea is: For now ContainerOp needs to have a unique ID (due to output PipelineParams serialization to string). It can either generate the ID by itself or get the ID from an "external" system like the DSL compiler.

Reply:
Thanks for the explanation. I missed that the add_op generates a unique name.

Reply:
I agree that the `add_op` name is somewhat confusing as it does not mention some of the method's behavior.
We should probably clean this place in future.